mullvad-relay-selector/src/relay_selector/query.rs

```diff
@@ -0,0 +1,779 @@
+//!
+//! The builder patterns included in the module simplify the process of constructing
+//! queries and ensure that queries are built in a type-safe manner, reducing the risk
+//! of runtime errors and improving code readability.
+
+use mullvad_types::{
+    constraints::Constraint,
+    relay_constraints::{
+        BridgeConstraints, LocationConstraint, OpenVpnConstraints, Ownership, Providers,
+        RelayConstraints, SelectedObfuscation, TransportPort, Udp2TcpObfuscationSettings,
+        WireguardConstraints,
+    },
+};
+use talpid_types::net::{proxy::CustomProxy, IpVersion, TunnelType};
+
+/// Represents a query for a relay based on various constraints.
+///
+/// This struct contains constraints for the location, providers, ownership,
+/// tunnel protocol, and additional protocol-specific constraints for WireGuard
+/// and OpenVPN. These constraints are used by the [`crate::RelaySelector`] to
+/// filter and select suitable relay servers that match the specified criteria.
+///
+/// A [`RelayQuery`] is best constructed via the fluent builder API exposed by
+/// [`builder::RelayQueryBuilder`].
+///
+/// # Examples
+///
+/// Creating a basic `RelayQuery` to filter relays by location, ownership and tunnel protocol:
+///
+/// ```rust
+/// // Create a query for a Wireguard relay that is owned by Mullvad and located in Norway.
+/// // The endpoint should specify port 443.
+/// use mullvad_relay_selector::query::RelayQuery;
+/// use mullvad_relay_selector::query::builder::RelayQueryBuilder;
+/// use mullvad_relay_selector::query::builder::{Ownership, GeographicLocationConstraint};
+///
+/// let query: RelayQuery = RelayQueryBuilder::new()
+///     .wireguard()                                            // Specify the tunnel protocol
+///     .location(GeographicLocationConstraint::country("no"))  // Specify the country as Norway
+///     .ownership(Ownership::MullvadOwned)                     // Specify that the relay must be owned by Mullvad
+///     .port(443)                                              // Specify the port to use when connecting to the relay
+///     .build();                                               // Construct the query
+/// ```
+///
+/// This example demonstrates creating a `RelayQuery` which can then be passed
+/// to the [`crate::RelaySelector`] to find a relay that matches the criteria.
+/// See [`builder`] for more info on how to construct queries.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RelayQuery {
+    pub location: Constraint<LocationConstraint>,
+    pub providers: Constraint<Providers>,
+    pub ownership: Constraint<Ownership>,
+    pub tunnel_protocol: Constraint<TunnelType>,
+    pub wireguard_constraints: WireguardRelayQuery,
+    pub openvpn_constraints: OpenVpnRelayQuery,
+}
+
+impl RelayQuery {
+    /// Create a new [`RelayQuery`] with no opinionated defaults. This
+    /// should be the const equivalent to [`Default::default`].
```

Comment:
```suggestion
    /// Create a new [`RelayQuery`] with no opinionated defaults. This query matches every relay
    /// with any configuration by setting each of its fields to [`Constraint::Any`]. Should be the const
    /// equivalent to [`Default::default`].
```

Reply:
Even better:
```suggestion
    /// Create a new [`RelayQuery`] with no opinionated defaults. This query matches every relay
    /// with any configuration by setting each of its fields to [`Constraint::Any`]. Should be the
    /// const equivalent to [`Default::default`].
    ///
    /// Note that the following identity applies for any `other_query`:
    /// ```rust
    /// # use mullvad_relay_selector::query::RelayQuery;
    /// # use crate::mullvad_relay_selector::query::Intersection;
    /// # let other_query = RelayQuery::new();
    /// assert_eq!(RelayQuery::new().intersection(other_query.clone()), Some(other_query));
    /// # let other_query = RelayQuery::new();
    /// assert_eq!(other_query.clone().intersection(RelayQuery::new()), Some(other_query));
    /// ```
```

Reply:
Nice!